lib/Dialect/AIG/Transforms/GreedyCutDecomp.cpp

```diff
@@ -0,0 +1,292 @@
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+struct AndInverterOpToCutPattern : public OpRewritePattern<aig::AndInverterOp> {
+  using OpRewritePattern<aig::AndInverterOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(aig::AndInverterOp op,
+                                PatternRewriter &rewriter) const override {
+    if (isa<aig::CutOp>(op->getParentOp()))
+      return failure();
+
+    auto cutOp = rewriter.create<aig::CutOp>(
+        op.getLoc(), op.getResult().getType(), op.getInputs(), [&]() {
+          auto result = rewriter.create<aig::AndInverterOp>(
+              op.getLoc(), op.getResult().getType(),
+              rewriter.getBlock()->getArguments(), op.getInvertedAttr());
+          rewriter.create<aig::OutputOp>(op.getLoc(), ValueRange{result});
+        });
+
+    rewriter.replaceOp(op, cutOp);
+    return success();
+  }
+};
+
+static aig::CutOp mergeCuts(Location loc, MutableArrayRef<Operation *> cuts,
+                            ArrayRef<Value> inputs, Value output,
+                            PatternRewriter &rewriter) {
+  if (!mlir::computeTopologicalSorting(cuts))
+    return {};
+
+  assert(cuts.size() >= 2);
+
+  DenseMap<Value, Value> valueToNewValue, inputsToBlockArg;
+  auto cutOp =
+      rewriter.create<aig::CutOp>(loc, output.getType(), inputs, [&]() {
+        for (auto [i, input] : llvm::enumerate(inputs))
+          inputsToBlockArg[input] = rewriter.getBlock()->getArgument(i);
+
+        for (auto [i, cut] : llvm::enumerate(cuts)) {
+          auto cutOp = cast<aig::CutOp>(cut);
+          assert(cutOp.getNumResults() == 1);
+          for (auto [arg, input] : llvm::zip(
+                   cutOp.getBodyBlock()->getArguments(), cutOp.getInputs())) {
+            auto it = inputsToBlockArg.find(input);
+            if (it != inputsToBlockArg.end()) {
+              rewriter.replaceAllUsesWith(arg, it->second);
+            } else {
+              auto cutOp = dyn_cast<aig::CutOp>(input.getDefiningOp());
+              assert(cutOp && cutOp.getNumResults() == 1);
+              rewriter.replaceAllUsesWith(arg, valueToNewValue.at(input));
+            }
+          }
+
+          assert(cutOp.getNumResults() == 1);
+          valueToNewValue[cutOp.getResult(0)] =
+              cutOp.getBodyBlock()->getTerminator()->getOperand(0);
+        }
+      });
+
+  rewriter.replaceAllUsesWith(output, cutOp.getResult(0));
+  {
+    OpBuilder::InsertionGuard guard(rewriter);
+    rewriter.setInsertionPointToEnd(cutOp.getBodyBlock());
+    rewriter.create<aig::OutputOp>(loc, ValueRange{valueToNewValue.at(output)});
+  }
+
+  for (auto oldCut : llvm::reverse(cuts)) {
+    auto *oldCutBlock = cast<aig::CutOp>(oldCut).getBodyBlock();
+    auto oldCutOutput = oldCutBlock->getTerminator();
+    oldCutOutput->erase();
+    // Erase arguments before inlining. Arguments are already replaced.
+    oldCutBlock->eraseArguments([](BlockArgument block) { return true; });
```

Comment:
I think these things should go through the rewriter and not be done directly on the op/block.

Reply:
Thank you for pointing out. oldCutOutput must be erased through rewriter but block arguments should be fine to erase directly. There is no API in rewriter to erase block arguments and upstream does the same (https://github.com/llvm/llvm-project/blob/a285ba7529feaa7c49890e314facb5e9f4d8dc11/mlir/lib/Conversion/TosaToSCF/TosaToSCF.cpp#L38-L39)